src/VisualStudio/Core/Def/ProjectSystem/InvisibleEditor.cs

```diff
@@ -54,37 +54,60 @@ public InvisibleEditor(IServiceProvider serviceProvider, string filePath, IVsHie
+                    Marshal.ThrowExceptionForHR(((IVsUndoState)_manager).IsEnabled(out var isEnabled));
+                    _needsUndoRestored = isEnabled != 0;
+                    if (_needsUndoRestored)
                     {
-                        Marshal.ThrowExceptionForHR(_vsTextLines.GetUndoManager(out _manager));
-                        Marshal.ThrowExceptionForHR(((IVsUndoState)_manager).IsEnabled(out var isEnabled));
-                        _needsUndoRestored = isEnabled != 0;
-                        if (_needsUndoRestored)
-                        {
-                            _manager.DiscardFrom(null); // Discard the undo history for this document
-                            _manager.Enable(0); // Disable Undo for this document
-                        }
+                        _manager.DiscardFrom(null); // Discard the undo history for this document
+                        _manager.Enable(0); // Disable Undo for this document
                     }
                 }
-                finally
-                {
-                    Marshal.Release(docDataPtr);
-                }
             }
             finally
             {
                 // We need to clean up the extra reference we have, now that we have an RCW holding onto the object.
                 Marshal.Release(invisibleEditorPtr);
             }
+
+            // Try casting the doc data to IVsTextLines first.
+            // If it fails try casting to IVsTextBufferProvider as some files like .aspx use that to provide the buffer
+            static IVsTextLines RetrieveDocData(IVsInvisibleEditor invisibleEditor, bool needsSave)
+            {
+                IVsTextLines? buffer = null;
+
+                var hr = invisibleEditor.GetDocData(fEnsureWritable: needsSave ? 1 : 0, riid: typeof(IVsTextLines).GUID, ppDocData: out var docDataPtrViaTextLines);
+                if (ErrorHandler.Succeeded(hr) &&
+                    Marshal.GetObjectForIUnknown(docDataPtrViaTextLines) is IVsTextLines vsTextLines)
+                {
+                    buffer = vsTextLines;
+                }
+                else
+                {
+                    hr = invisibleEditor.GetDocData(fEnsureWritable: needsSave ? 1 : 0, riid: typeof(IVsTextBufferProvider).GUID, ppDocData: out var docDataPtrViaTextBufferProvider);
+                    if (ErrorHandler.Succeeded(hr) &&
+                        Marshal.GetObjectForIUnknown(docDataPtrViaTextBufferProvider) is IVsTextBufferProvider vsTextBufferProvider)
+                    {
+                        hr = vsTextBufferProvider.GetTextBuffer(out buffer);
+                    }
+
+                    if (docDataPtrViaTextBufferProvider != IntPtr.Zero)
+                        Marshal.Release(docDataPtrViaTextBufferProvider);
+                }
+
+                if (docDataPtrViaTextLines != IntPtr.Zero)
+                    Marshal.Release(docDataPtrViaTextLines);
```

Comment:
Do you want to put this in a try/finally just to really ensure there's no chance it gets skipped? My concern would be if there was an IVsTextBufferProvider for example that was implemented in managed code and that GetTextBuffer call throws.

Reply:
That would be evil, but probably better to fight the evil than run away from it.

Reply:
The other option is to use something like https://learn.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.errorhandler.callwithcomconvention but this also protects against our own mistakes too.